### ISSUE TITLE
docker doesn't unpack local tarball with file file capabilities/xattr when using ADD in Dockerfile

### ISSUE THREAD

**Issue by jperrin** [docker/docker#20331]
When creating a base container, Docker does not unpack a local tarball as expected when using the ADD directive if the tarball contains files with capabilities referenced in xattr, but instead includes the tarball whole in the container. 

I have created two tarballs that demonstrate this behavior, located at http://people.centos.org/jperrin/tars/

The bad behavior can be seen when using those two tarballs with a very simple Dockerfile...

```
FROM scratch
MAINTAINER me
ADD replace-with-tarball-name /
CMD ["/bin/bash"]
```

Both builds will complete, however one will be roughly 197MB in size, while the other will be 42MB or so. If you examine the contents of the smaller container via `docker export | tar -t ` you'll see that the tarball is included rather than unpacked. 

This breakage appears to be the result of the file capabilities on `/usr/bin/ping` and a few other binaries inside the tarball. 

The tarball that unpacks successfully was created with `tar -C /mnt/ -Jcf centos-img.tar.xz .` which does not preserve xattr info by default. 

The tarball that fails to be properly unpacked via the ADD was created with `tar --xattrs -C /mnt/ -Jcf centos-img.tar.xz .`

**Comment by thaJeztah** [comment]
ping @vbatts I think you looked into xattrs on some other occasions?

**Comment by vbatts** [comment]
```bash
$ docker version
Client:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Mon Feb 15 10:55:21 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Mon Feb 15 10:55:21 2016
 OS/Arch:      linux/amd64
$ touch file
$ sudo setfattr -n security.capability -v farts file
$ tar --xattrs -cf f.tar ./file
$ strings ./f.tar | grep SCHILY
42 SCHILY.xattr.security.capability=farts
71 SCHILY.xattr.security.selinux=unconfined_u:object_r:unlabeled_t:s0
$ cat Dockerfile
FROM fedora
RUN dnf install -y attr
ADD f.tar /
RUN attr -l file
$ docker build  .
Sending build context to Docker daemon 15.36 kB
Step 1 : FROM fedora
 ---> 6547ce9b3407
Step 2 : RUN dnf install -y attr
 ---> Using cache
 ---> 93be2eb6f7aa
Step 3 : ADD f.tar /
 ---> 1e95d939a367
Removing intermediate container aadc06378a4c
Step 4 : RUN attr -l file
 ---> Running in 9245fa46f994
Attribute "capability" has a 5 byte value for file
Attribute "selinux" has a 37 byte value for file
 ---> a1513b446648
Removing intermediate container 9245fa46f994
Successfully built a1513b446648
```

I'll have to double check further what it is about the xattrs from @jperrin archives that are not jiving, because in this most basic test it works.

**Comment by jperrin** [comment]
interesting. I see this behavior with both 1.9.1 on f23, and 1.8.2 on el7. I'll try with your method above and see if I can't narrow it down a bit more.

**Comment by unclejack** [comment]
@jperrin Have you had a chance to look any further into this?

**Comment by jperrin** [comment]
I haven't, unfortunately. I'm in the middle of a two week travel window, so likely won't be able to look at it until next week.

**Comment by rhatdan** [comment]
Is the problem here, not xattrs but setting of file capabilties?  Perhaps this requires SYS_ADMIN capability?

**Comment by twaugh** [comment]
The actual syscall that seems to cause this failure:
```
[pid 20411] lsetxattr("/dev", "security.selinux", "system_u:object_r:root_t:s0", 27, 0) = -1 EOPNOTSUPP (Operation not supported)
```

**Comment by lcarva** [comment]
@vbatts to reproduce this issue a tarball with nested dirs/files is needed.

```bash
$ mkdir -p contents/var/foo
$ touch contents/var/foo/nested-file
$ attr -l contents/var/foo/nested-file 
Attribute "selinux" has a 37 byte value for contents/var/foo/nested-file
$ cd contents
$ tar --xattrs -cf ../f.tar .
$ cd ..
$ strings ./f.tar | grep SCHILY
71 SCHILY.xattr.security.selinux=unconfined_u:object_r:home_root_t:s0
71 SCHILY.xattr.security.selinux=unconfined_u:object_r:home_root_t:s0
71 SCHILY.xattr.security.selinux=unconfined_u:object_r:home_root_t:s0
71 SCHILY.xattr.security.selinux=unconfined_u:object_r:home_root_t:s0
$ cat Dockerfile
FROM fedora
ADD f.tar /
RUN ls /var/foo
$ docker build --no-cache .                                                                                       
Sending build context to Docker daemon   510 MB
Step 1 : FROM fedora
 ---> 9bdb5101e5fc
Step 2 : ADD f.tar /
 ---> 85c1f4cf0875
Removing intermediate container 2612144cfd46
Step 3 : RUN ls /var/foo
 ---> Running in 9e7b6eb7b214
ls: cannot access /var/foo: No such file or directory
The command '/bin/sh -c ls /var/foo' returned a non-zero code: 2

```

When omitting `--xattrs` during tarball creation, the `ls` command works properly in the container.

**Comment by rhatdan** [comment]
We should not be storing SELinux Xattrs here in the first place.  They have no reason to be saved.

**Comment by rhatdan** [comment]
The stuff I would worry about is file capabilities.

**Comment by vbatts** [comment]
f23

```

vbatts@f ~/x (master) $ cd contents/
vbatts@f ~/x/contents (master) $ tar --xattrs -cvf ../f.tar  .
./
./var/
./var/foo/
./var/foo/nested-file
vbatts@f ~/x/contents (master) $ b
vbatts@f ~/x (master) $ docker build --no-cache .
Sending build context to Docker daemon 14.85 kB
Step 1 : FROM fedora
 ---> ddd5c9c1d0f2
Step 2 : ADD f.tar /
 ---> 92571a0ab5e7
Removing intermediate container 2d80828a8163
Step 3 : RUN ls /var/foo
 ---> Running in ee8b1cd1bf37
nested-file
 ---> f83189955163
Removing intermediate container ee8b1cd1bf37
Successfully built f83189955163

```

:-\

**Comment by maxamillion** [comment]
@vbatts try it with an image produced from the Fedora build system http://koji.fedoraproject.org/koji/taskinfo?taskID=13942066

**Comment by maxamillion** [comment]
I'm not entirely certain how to work around this and we need to get a Docker image update out for Fedora.

https://github.com/docker-library/official-images/issues/1691

**Comment by lcarva** [comment]
@vbatts Could be related to docker version:

```
$ docker version
Client:
 Version:         1.9.1
 API version:     1.21
 Package version: docker-1.9.1-9.gitee06d03.fc23.x86_64
 Go version:      go1.5.3
 Git commit:      ee06d03/1.9.1
 Built:           
 OS/Arch:         linux/amd64

Server:
 Version:         1.9.1
 API version:     1.21
 Package version: docker-1.9.1-9.gitee06d03.fc23.x86_64
 Go version:      go1.5.3
 Git commit:      ee06d03/1.9.1
 Built:           
 OS/Arch:         linux/amd64
```

I created a gist that might help others in checking this behavior: https://gist.github.com/lcarva/549f5569de85845209751867f374c457

**Comment by twaugh** [comment]
Is everyone who sees problems using the same storage driver?

`docker info | grep -1 'Storage Driver'`

**Comment by lcarva** [comment]
@twaugh Here's mine:

```bash
$ docker info | grep -1 'Storage Driver'
Server Version: 1.9.1
Storage Driver: devicemapper
 Pool Name: docker-253:1-925481-pool

```

**Comment by maxamillion** [comment]
@twaugh 

```
$ docker info | grep -1 'Storage Driver'
Server Version: 1.9.1
Storage Driver: devicemapper
 Pool Name: docker-253:1-136225-pool
```

**Comment by twaugh** [comment]
I have two systems here, one using devicemapper with /dev/loop* and the other using direct-lvm with an LVM thin pool allocated for it. I don't see this problem with devicemapper+direct-lvm.

**Comment by twaugh** [comment]
@maxamillion, @lcarva, can either (or both) of you try setting up a new machine, maybe a VM, with devicemapper+direct-lvm? On Fedora 23 I just put "STORAGE_DRIVER=devicemapper" and "VG=docker" in /etc/sysconfig/docker-storage-setup, used pvcreate on a spare partition, then "vgcreate docker /dev/..." to create the volume group, and finally run docker-storage-setup to sort out the thin pool. Do this before starting docker for the first time.

After doing that, can you still reproduce the problem?

**Comment by maxamillion** [comment]
```
# cat /etc/sysconfig/docker-storage
# This file may be automatically generated by an installation program.
# Please DO NOT edit this file directly. Instead edit
# /etc/sysconfig/docker-storage-setup and/or refer to
# "man docker-storage-setup".

# By default, Docker uses a loopback-mounted sparse file in
# /var/lib/docker.  The loopback makes it slower, and there are some
# restrictive defaults, such as 100GB max storage.

DOCKER_STORAGE_OPTIONS="--storage-driver=devicemapper --storage-opt dm.datadev=/dev/vg-docker/data --storage-opt dm.metadatadev=/dev/vg-docker/metadata"

# docker info | grep -1 'Storage Driver'
Server Version: 1.9.1
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Storage Driver: devicemapper
 Pool Name: docker-253:0-1115764-pool

# ps -ef | grep 'docker daemon'
root     10166     1  0 21:29 ?        00:00:00 /bin/sh -c /usr/bin/docker daemon $OPTIONS            $DOCKER_STORAGE_OPTIONS            $DOCKER_NETWORK_OPTIONS            $ADD_REGISTRY            $BLOCK_REGISTRY            $INSECURE_REGISTRY            2>&1 | /usr/bin/forward-journald -tag docker
root     10167 10166  0 21:29 ?        00:00:00 /usr/bin/docker daemon --selinux-enabled --storage-driver=devicemapper --storage-opt dm.datadev=/dev/vg-docker/data --storage-opt dm.metadatadev=/dev/vg-docker/metadata
root     10314  4738  0 21:31 pts/0    00:00:00 grep --color=auto docker daemon

# docker build -t fedoratest .
Sending build context to Docker daemon  44.8 MB
Step 1 : FROM scratch
 ---> 
Step 2 : MAINTAINER Adam Miller <maxamillion@fedoraproject.org>
 ---> Running in 79489dbf2d12
 ---> 64c9adbf901d
Removing intermediate container 79489dbf2d12
Step 3 : ADD fedora-23-20160506.tar.xz /
 ---> b2b37907ede4
Removing intermediate container 249279a7ede3
Successfully built b2b37907ede4

# docker run -ti fedoratest /bin/bash
exec: "/bin/bash": stat /bin/bash: no such file or directory
Error response from daemon: Cannot start container 0d67d46c826dd52957b15c2b0518e26f1abf12b32d15bc1aa3535e7c1f7a8a41: [8] System error: exec: "/bin/bash": stat /bin/bash: no such file or directory

**Comment by twaugh** [comment]
OK. Here's `docker info` from the system that doesn't show the problem here:
https://gist.github.com/twaugh/7c64ed28f5c598fa455c5ace7fca5534

**Comment by imcleod** [comment]
For informational purposes, here is the actual tar command that generates the output from koji/Factory:

https://github.com/redhat-imaging/imagefactory/commit/bba65473918d31f3c7330383b71d304466883cca#diff-8dc103de1c457ee19ae2309520550af4R205

In other words:

tar -cf <output_file> -C <input_filesystem_dir> --acls --xattrs ./

I explicitly added --acls and --xattrs in response to some bugs raised against our base images related to SELinux.

@rhatdan or @vbatts  - is there a better/different set of parameters I should be using here to avoid the issues described above?

**Comment by rhatdan** [comment]
Well there is no reason for saving the SELinux labels in the tar format, those are totally useless and most likely misleading.  Not sure if their is any other --xttrs of any value.  
The --acls might be of some value and the file capabilities, but I am not sure how those are stored.

**Comment by macropin** [comment]
@jperrin Is this a blocker for https://github.com/docker-library/official-images/issues/1691?

**Comment by jperrin** [comment]
It's not helping matters, but I can work around it. We had to wait for
our upstream to release the sources for 6.8. We should have the builds
out today for 6.

On 05/15/2016 11:03 PM, Andrew Cutler wrote:
> @jperrin <https://github.com/jperrin> Is this a blocker for
> docker-library/official-images#1691
> <https://github.com/docker-library/official-images/issues/1691>?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/issues/20331#issuecomment-219344160>
> 

-- 
Jim Perrin
The CentOS Project | http://www.centos.org
twitter: @BitIntegrity | GPG Key: FA09AD77

**Comment by maxamillion** [comment]
FWIW this is still broken in docker 1.11 but broken differently.

```
# docker build -t fedora:24 . 
Sending build context to Docker daemon 43.83 MB
Step 1 : FROM scratch
 ---> 
Step 2 : MAINTAINER Adam Miller <maxamillion@fedoraproject.org>
 ---> Running in dec5e12a6c5c
 ---> c020593e26e9
Removing intermediate container dec5e12a6c5c
Step 3 : ADD fedora-24-20160525.tar.xz /
Untar re-exec error: exit status 1: output: operation not supported
```

**Comment by vbatts** [comment]
i'm working to narrow this reproducer, but this issue is less than clear. I hate that it's a blocker. I've just attempted this on a tar archive with only the `ping` binary, and it works fine. See the output below. This is on fedora 23.
```bash
vbatts@f ~ (master *) $ docker version
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Tue Mar 22 19:18:48 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Tue Mar 22 19:18:48 2016
 OS/Arch:      linux/amd64
vbatts@f ~ (master *) $ tar -cf ~/x/out.tar --xattrs --acls /usr/bin/ping
tar: Removing leading `/' from member names
vbatts@f ~ (master *) $ cd x
vbatts@f ~/x (master *) $ cat Dockerfile 
FROM scratch
ADD out.tar /
vbatts@f ~/x (master *) $ docker build -t gh20331 .
Sending build context to Docker daemon 53.76 kB
Step 1 : FROM scratch
 ---> 
Step 2 : ADD out.tar /
 ---> 6165791e0e1d
Removing intermediate container 8b7e7003e5df
Successfully built 6165791e0e1d
vbatts@f ~/x (master *) $ docker save gh20331 | tar xv
6165791e0e1dc164853a54f3d5f6ae83944fb6fd71364d9c53768cc15664fa8f.json
cc81937da72d181903b2077055305e75cd238a40df48bc3a00ca22b30e876ad4/
cc81937da72d181903b2077055305e75cd238a40df48bc3a00ca22b30e876ad4/VERSION
cc81937da72d181903b2077055305e75cd238a40df48bc3a00ca22b30e876ad4/json
cc81937da72d181903b2077055305e75cd238a40df48bc3a00ca22b30e876ad4/layer.tar
manifest.json
repositories
vbatts@f ~/x (master *) $ tar --xattrs-include='*' -tvvvf cc81937da72d181903b2077055305e75cd238a40df48bc3a00ca22b30e876ad4/layer.tar
drwxr-xr-x  0/0               0 2016-05-31 15:35 usr/
drwxr-xr-x  0/0               0 2016-05-31 15:35 usr/bin/
-rwxr-xr-x* 0/0           44752 2015-11-19 15:14 usr/bin/ping
  x: 20 security.capability
vbatts@f ~/x (master *) $ tar --xattrs-include='*' -xvvvf cc81937da72d181903b2077055305e75cd238a40df48bc3a00ca22b30e876ad4/layer.tar
drwxr-xr-x  0/0               0 2016-05-31 15:35 usr/
drwxr-xr-x  0/0               0 2016-05-31 15:35 usr/bin/
-rwxr-xr-x* 0/0           44752 2015-11-19 15:14 usr/bin/ping
  x: 20 security.capability
tar: setxattrat: Cannot set 'security.capability' extended attribute for file 'usr/bin/ping': Operation not permitted
vbatts@f ~/x (master *) $ sudo !!
sudo tar --xattrs-include='*' -xvvvf cc81937da72d181903b2077055305e75cd238a40df48bc3a00ca22b30e876ad4/layer.tar
drwxr-xr-x  0/0               0 2016-05-31 15:35 usr/
drwxr-xr-x  0/0               0 2016-05-31 15:35 usr/bin/
-rwxr-xr-x* 0/0           44752 2015-11-19 15:14 usr/bin/ping
  x: 20 security.capability
vbatts@f ~/x (master *) $ getcap usr/bin/ping
usr/bin/ping = cap_net_admin,cap_net_raw+ep
```

**Comment by maxamillion** [comment]
@vbatts Anything I can provide here that might be helpful? Did you attempt https://github.com/docker/docker/issues/20331#issuecomment-217563282 ?

**Comment by justincormack** [comment]
I am looking at this as well, trying to replicate. I think it may be an issue with some storage drivers.

**Comment by vbatts** [comment]
@maxamillion sorry. I had not, and now I get a permission denied on pulling that rootfs

**Comment by vbatts** [comment]
@justincormack I was on overlay with 4.4.9-300.fc23.x86_64 kernel

**Comment by justincormack** [comment]
The original issue I can replicate on 1.9.1.fc23, but is fixed on 1.11.

@maxamillion can you point me at a URL with the fedora archive that breaks for you on 1.11? I can't replicate it with any of the images that are listed above that I can get hold of.

Note that extended attributes are not supported in docker images and should be ignored anyway (although you can set them at runtime generally). See #1070 #339, the issue is image portability. This does potentially lead to some issues around ability for non root users to access binaries that need additional capabilities, generally `ping` although this mainly affects Fedora and I believe you are moving to ping sockets which should work although I may add some support to make that easier.

**Comment by maxamillion** [comment]
@vbatts @justincormack Any of the `Fedora-Docker-Base-24`  tasks [here](http://koji.fedoraproject.org/koji/tasks?start=0&state=all&view=tree&method=image&order=-id) should suffice.

Here is a recent one

 https://kojipkgs.fedoraproject.org//work/tasks/111/14350111/Fedora-Docker-Base-24-20160602.n.0.x86_64.tar.xz

**Comment by vbatts** [comment]
@justincormack IIRC the biggest blockers were around AUFS lacking xattr support. Oddly, the image was still portable with xattrs and AUFS, but just unable to get/set inside a container.

**Comment by justincormack** [comment]
@maxamillion that image works fine for me under 1.11 (official binary via get.docker.com on fc23) using
```
FROM scratch
ADD 87fab38d658cbbd99872ce3c817dd7e11d9274916dd9f40601b91ca0c9bfb1c5/layer.tar .
```
The image doesn't contain `ping` though.

@vbatts recent aufs does have xattr support as an option now, and it works in `docker run` fine, but I think the concerns are around user xattr still being patchily supported, and that you cannot round trip to eg Windows or BSD, and now layer checksums would need to support attributes that are not portable as well.

The main issue is currently for eg #8460 where users cannot actually add filesystem capabilities for non root use, so I think adding ambient capabilities may be the best solution.

**Comment by maxamillion** [comment]
@justincormack when I tested it a week ago it didn't work with devicemapper, what storage driver are you using?

**Comment by vbatts** [comment]
@justincormack well windows and bsd it doens't entirely apply (though FreeBSD does have extattr support ...). Besides, if those platforms can include the tar-split support, then it is a moot point as well.

**Comment by vbatts** [comment]
Docker v1.10.3 on slackware with btrfs backend 
```shell
vbatts@bananaboat ~/x (master *) $ docker build --no-cache -t gh20331 .
Sending build context to Docker daemon 44.99 MB
Step 1 : FROM scratch
 ---> 
Step 2 : ADD Fedora-Docker-Base-24-20160602.n.0.x86_64.tar.xz /
 ---> f16387627950
Removing intermediate container 4e4c64729a14
Successfully built f16387627950
vbatts@bananaboat ~/x (master *) $ docker save gh20331 | tar xO 8228e1084874abfb3b8e389c9e26f5155c8c2a4dc910188e8ab927f12ea7e755/layer.tar | strings | grep SCHILY | sort -u   
64 SCHILY.xattr.security.selinux=system_u:object_r:fusefs_t:s0
93 SCHILY.xattr.system.posix_acl_access=
94 SCHILY.acl.access=user::rwx
94 SCHILY.xattr.system.posix_acl_default=
95 SCHILY.acl.default=user::rwx
Can't allocate memory for SCHILY.acl.access
Can't allocate memory for SCHILY.acl.default
Parse error: SCHILY.acl.access
Parse error: SCHILY.acl.default
SCHILY.acl.access
SCHILY.acl.default
SCHILY.dev
SCHILY.devmajor
SCHILY.devminor
SCHILY.fflags
SCHILY.ino
SCHILY.nlink
SCHILY.realsize
```

Docker v1.10.3 on f23 and overlay backend
```shell
vbatts@f ~/gh20331 (master *) $ docker build --no-cache -t gh20331 .                                                                         
Sending build context to Docker daemon 44.98 MB                                                                                              
Step 1 : FROM scratch                                                                                                                        
 --->                                                                                                                                        
Step 2 : ADD Fedora-Docker-Base-24-20160602.n.0.x86_64.tar.xz /                                                                              
 ---> 79cf9ac2206f                                                                                                                           
Removing intermediate container 90c07a398493                                                                                                 
Successfully built 79cf9ac2206f                                                                                                              
vbatts@f ~/gh20331 (master *) $ docker save gh20331 | tar tv | grep layer                                                                    
-rw-r--r-- 0/0       226299392 2016-06-02 14:54 ca21d5d92876e69d7b05ec077d2ba7566a261d6934eaaa50e49bb5cc389a12d7/layer.tar                   
vbatts@f ~/gh20331 (master *) $ docker save gh20331 | tar xO ca21d5d92876e69d7b05ec077d2ba7566a261d6934eaaa50e49bb5cc389a12d7/layer.tar | strings | grep SCHILY | sort -u                                                                                                                 
64 SCHILY.xattr.security.selinux=system_u:object_r:fusefs_t:s0                                                                               
93 SCHILY.xattr.system.posix_acl_access=                                                                                                     
94 SCHILY.acl.access=user::rwx                                                                                                               
94 SCHILY.xattr.system.posix_acl_default=                                                                                                    
95 SCHILY.acl.default=user::rwx                                                                                                              
Can't allocate memory for SCHILY.acl.access                                                                                                  
Can't allocate memory for SCHILY.acl.default                                                                                                 
Parse error: SCHILY.acl.access                                                                                                               
Parse error: SCHILY.acl.default                                                                                                              
SCHILY.acl.access                                                                                                                            
SCHILY.acl.default                                                                                                                           
SCHILY.dev                                                                                                                                   
SCHILY.devmajor                                                                                                                              
SCHILY.devminor                                                                                                                              
SCHILY.fflags                                                                                                                                
SCHILY.ino                                                                                                                                   
SCHILY.nlink                                                                                                                                 
SCHILY.realsize
```

**Comment by justincormack** [comment]
@vbatts confused, there you are never untaring the tarball with the acls in anyway, as it is inside the `Fedora-Docker-Base-24-20160602.n.0.x86_64.tar.xz` tarball?

**Comment by justincormack** [comment]
@vbatts the problem is the attributes are not roundtrippable between different systems even if they have support, due to different size and name limits and so on.

**Comment by vbatts** [comment]
@justincormack the untarring seems to be a separate issue. Traditionally `ADD Fedora-Docker-Base-24-20160602.n.0.x86_64.tar.xz /` would extract to `/` but it now seems that it just places the tar archive at `/` ... 🙍

**Comment by vbatts** [comment]
@justincormack I understand those pieces. That is one of the impetus of tar-split, to preserve that order regardless of how the host/platform mangles it.

**Comment by justincormack** [comment]
@vbatts it works for me, eg testing with
```
FROM alpine
RUN mkdir /out
WORKDIR /out
ADD Fedora-Docker-Base-24-20160602.n.0.x86_64.tar.xz /out
```
then
```
docker run 0b4fbb0cb825 ls
87fab38d658cbbd99872ce3c817dd7e11d9274916dd9f40601b91ca0c9bfb1c5
repositories
```

You need to have `xz` installed on the host, but thats normally there.

**Comment by justincormack** [comment]
@maxamillion I am using `devicemapper` (loopback, but I don't think that should make a difference).

**Comment by maxamillion** [comment]
@vbatts using that docker base image I linked to previously, trying to build on a rawhide machine.

```
# docker build -t fedtest:24 .
Sending build context to Docker daemon 132.8 MB
Step 1 : FROM scratch
 --->
Step 2 : MAINTAINER Adam Miller <maxamillion@fedoraproject.org>
 ---> Running in 2d559e78cc3e
 ---> dd64eef9ba1f
Removing intermediate container 2d559e78cc3e
Step 3 : ADD fedora-24-20160602.tar.xz /
Untar re-exec error: exit status 1: output: operation not supported

# rpm -q docker
docker-1.11.1-5.git9dea74f.fc25.x86_64

# docker -v
Docker version 1.11.1, build 9dea74f/1.11.1
[root@rawhide f24]# docker info
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 1
Server Version: 1.11.1
Storage Driver: devicemapper
 Pool Name: docker-253:0-1445364-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 17.24 MB
 Data Space Total: 107.4 GB
 Data Space Available: 36.18 GB
 Metadata Space Used: 589.8 kB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.147 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.125 (2016-05-14)
Logging Driver: journald
Cgroup Driver: systemd
Plugins:
 Volume: local
 Network: host bridge null
Kernel Version: 4.7.0-0.rc0.git10.1.fc25.x86_64
Operating System: Fedora 25 (Rawhide)
OSType: linux
Architecture: x86_64
Number of Docker Hooks: 2
CPUs: 2
Total Memory: 7.776 GiB
Name: rawhide
ID: 3Y2K:7RH6:DWK4:6J3N:6L5C:BC4K:GSSC:OGQQ:3F5M:H4RG:Z7FZ:WHKM
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
Registries: docker.io (secure)
```

**Comment by justincormack** [comment]
@maxamillion does the version of `tar` on rawhide do something different, eg extract extended attributes by default without any extra flags? Also can you try it with the official docker binary not the native rawhide one, as something is clearly different, trying to work out what...

**Comment by maxamillion** [comment]
@justincormack I tried but the docker official repo is failing.

```
# curl -fsSL https://get.docker.com/ | sh
+ sh -c 'sleep 3; dnf -y -q install docker-engine'
Failed to synchronize cache for repo 'docker-main-repo', disabling.
```

**Comment by justincormack** [comment]
Ah, you can use the static binaries (but they don't support devicemapper), or just build from source yourself.

**Comment by maxamillion** [comment]
@justincormack I built from source and confirmed that this works

```
# docker build -t fedora:24 .
Sending build context to Docker daemon 440.3 MB
Step 1 : FROM scratch
 --->
Step 2 : MAINTAINER Adam Miller <maxamillion@fedoraproject.org>
 ---> Using cache
 ---> 4d93eb60c915
Step 3 : ADD fedora-rawhide-20160602.tar.xz /
 ---> fd41da8714de
Removing intermediate container f9a1d9cd40b4
Successfully built fd41da8714de

# docker run --rm -ti fedora:24 /bin/bash
[root@9ce764f4d174 /]# exit
exit
```

**Comment by justincormack** [comment]
As this appears to be an issue with the Red Hat patched build I am going to close this now, but feel free to continue posting on this thread, especially as I cannot immediately see which patch would cause this and it is possible it is a mix of factors.

**Comment by runcom** [comment]
Reopening this as it appears the the `--selinux-enabled` flag is causing this issue even with the official Docker binaries; I tested it with 1.11.2 from Docker official from the install.sh script; adding the selinux flag makes ADD fail:
```
[root@rhel01 ~]# docker build --no-cache .
Sending build context to Docker daemon 427.4 MB
Step 1 : FROM scratch
 ---> 
Step 2 : ADD fedora-24-20160602.tar.xz /
Untar re-exec error: exit status 1: output: operation not supported
```

**Comment by runcom** [comment]
going to bisect this - I'll try with 1.9 just to be sure before 

@rhatdan I don't see any AVS, `setenforce 0` or `setenforce 1` doesn't make any difference here - it's the `--selinux-enabled` flag that's causing this

**Comment by justincormack** [comment]
Hmm that is odd.

**Comment by justincormack** [comment]
If you get a chance can you test on master as well?

**Comment by runcom** [comment]
On it

**Comment by runcom** [comment]
This is basically always reproducible on a Fedora or RHEL machine even with upstream binaries by setup dockerd to use --selinux-enabled

**Comment by runcom** [comment]
on a RHEL 7.2 with Docker official master this is fixed (it shows `-unsupported` because I had to tweak something to compile with go1.4.2)
```
[root@rhel01 ~]# docker version
Client:
 Version:      1.12.0-dev
 API version:  1.24
 Go version:   go1.4.2
 Git commit:   d85491f-unsupported
 Built:        Wed Jun  8 04:14:31 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-dev
 API version:  1.24
 Go version:   go1.4.2
 Git commit:   d85491f-unsupported
 Built:        Wed Jun  8 04:14:31 2016
 OS/Arch:      linux/amd64
```

**Comment by runcom** [comment]
The following commit fixed this on master:
```
[root@rhel01 docker-git]# git bisect bad
75a271603c761c18c7deb186d08943dcf85b739d is the first bad commit
commit 75a271603c761c18c7deb186d08943dcf85b739d
Author: Aleksa Sarai <asarai@suse.de>
Date:   Wed Mar 16 22:54:57 2016 +1100

    pkg: archive: don't fail Untar if xattrs are not supported
    
    Since certain filesystems don't support extended attributes, ignore
    errors produced (emitting a warning) when attempting to apply extended
    attributes to file.
    
    Signed-off-by: Aleksa Sarai <asarai@suse.de>

:040000 040000 c1c549656c03116abf335bb70346e35bcfe43fe6 56f1cbf51bed8b6603c543f337bbab70034598fd M	pkg
```
Let's leave this issue open though because it's related to https://github.com/docker/docker/issues/21216 which will have follow up PR iinw

**Comment by runcom** [comment]
I'm still not sure as to why docker+devmapper(xfs)+selinux-enabled does not support xattrs

**Comment by runcom** [comment]
smaller reproducer using docker 1.10.3 or 1.11.2 or anything before 75a271603c761c18c7deb186d08943dcf85b739d with `--selinux-enabled` (and devmapper):
```
[root@rhel01 repro]# tar -cf out.tar --xattrs --acls /usr/bin/ping
tar: Removing leading `/' from member names
[root@rhel01 repro]# cat > Dockerfile <<-EOF
> FROM scratch
> ADD out.tar /
> EOF
[root@rhel01 repro]# docker build .
Sending build context to Docker daemon 53.76 kB
Step 1 : FROM scratch
 ---> 
Step 2 : ADD out.tar /
[  430.838956] XFS (dm-1): Mounting V4 Filesystem
[  430.853515] XFS (dm-1): Ending clean mount
[  430.881924] XFS (dm-1): Unmounting Filesystem
[  430.965090] XFS (dm-1): Mounting V4 Filesystem
[  430.970844] XFS (dm-1): Ending clean mount
[  430.993924] XFS (dm-1): Unmounting Filesystem
[  431.061884] XFS (dm-1): Mounting V4 Filesystem
[  431.065126] XFS (dm-1): Ending clean mount
[  431.120207] XFS (dm-1): Unmounting Filesystem
Untar re-exec error: exit status 1: output: operation not supported
```

**Comment by runcom** [comment]
stracing the above yields:
```
[pid  3188] lsetxattr("/usr/bin/ping", "security.selinux", "system_u:object_r:ping_exec_t:s0", 33, 0) = -1 EOPNOTSUPP (Operation not supported)
```

**Comment by rhatdan** [comment]
The problem is that devmapper sets up a mount of the image as
```
 context="system_u:object_r:svirt_sandbox_file_t:s0:c1,c2"
```
When SELinux is enabled on devicemapper.  When a file system is mounted like this, attempt to change the SELinux label will be blocked.  Basically the kernel does not allow you to say label this mount point as XYZ, and allow a process to change a label of a file on the mount point to ABC.  The tar command should realize this and continue without complaint.  Actually the tar command should not even be attempting to set SELinux labels.  Since they will almost always be wrong.

**Comment by rhatdan** [comment]
```
[root@rhel01 docker-git]# git bisect bad
75a271603c761c18c7deb186d08943dcf85b739d is the first bad commit
commit 75a271603c761c18c7deb186d08943dcf85b739d
Author: Aleksa Sarai <asarai@suse.de>
Date:   Wed Mar 16 22:54:57 2016 +1100

    pkg: archive: don't fail Untar if xattrs are not supported

    Since certain filesystems don't support extended attributes, ignore
    errors produced (emitting a warning) when attempting to apply extended
    attributes to file.

    Signed-off-by: Aleksa Sarai <asarai@suse.de>
```
Is the correct patch.

**Comment by runcom** [comment]
Alright thanks @rhatdan

**Comment by tonistiigi** [comment]
@runcom @rhatdan So we can close this? Fixed in `v1.12`

**Comment by rhatdan** [comment]
I think so.

**Comment by justincormack** [comment]
Ok, closing then. Please open a new issue if there are any more issues with this.
